### PR TITLE
Adapt early stopping

### PR DESCRIPTION
**Pull request by daryadedik**
Adapting early stopping module (w/o unit tests).

### CODE REVIEW

File: expan/core/early_stopping.py [hunk 1]
```diff
@@ -1,73 +1,74 @@
...
+    """ Calculate an approximation of the O'Brien-Fleming alpha spending function.
 
-    Args:
-        information_fraction (scalar or array_like): share of the information 
-            amount at the point of evaluation, e.g. the share of the maximum 
-            sample size
-        alpha: type-I error rate
+    :param information_fraction: share of the information  amount at the point of evaluation, 
+                                 e.g. the share of the maximum sample size
+    :type  information_fraction: scalar or array-like
```
Comment (shansfolder) [inline expan/core/early_stopping.py]: float?

File: expan/core/early_stopping.py [hunk 2]
```diff
@@ -1,73 +1,74 @@
...
+
+    :return: results of type EarlyStoppingTestStatistics
+    :rtype:  EarlyStoppingTestStatistics
     """
     # Checking if data was provided
     if x is None or y is None:
         raise ValueError('Please provide two non-None samples.')
+    if not isinstance(x, pd.Series) or not isinstance(y, pd.Series):
+        if not isinstance(x, list) or not isinstance(y, list):
+            raise TypeError('Please provide samples of type Series or list.')
```
Comment (shansfolder) [inline expan/core/early_stopping.py]: Here we missed to check type ```np.ndarray```.
One way I did this check in the adopt statistics branch is this: https://github.com/zalando/expan/blob/adopt_statistics/expan/core/statistics.py#L59-L62

File: expan/core/early_stopping.py [hunk 3]
```diff
@@ -105,97 +106,63 @@ def group_sequential(x,
...
-            'treatment_mean'        : float(mu_x),
-            'control_mean'          : float(mu_y),
-            'treatment_variance'    : float(np.nanvar(_x)),
-            'control_variance'      : float(np.nanvar(_y))}
+                                       [alpha_new * 100 / 2, 100 - alpha_new * 100 / 2])
+
+    treatment_statistics = SampleStatistics(int(n_x), float(np.nanmean(_x)), float(np.nanvar(_x)))
+    control_statistics   = SampleStatistics(int(n_y), float(np.nanmean(_y)), float(np.nanvar(_y)))
+    variant_statistics   = BaseTestStatistics(control_statistics, treatment_statistics)
+    p_value              = statx.compute_p_value(_x, _y, int(n_x), int(n_y))
```
Comment (shansfolder) [inline expan/core/early_stopping.py]: maybe you already updated, but here we don't have parameters of sizes now.

File: expan/core/early_stopping.py [hunk 4]
```diff
@@ -316,40 +279,97 @@ def f(x, y):
...
-            'number_of_iterations'  : num_iters}
+    treatment_statistics = SampleStatistics(int(n_x), float(mu_x), float(np.nanvar(x)))
+    control_statistics   = SampleStatistics(int(n_y), float(mu_y), float(np.nanvar(y)))
+    variant_statistics   = BaseTestStatistics(control_statistics, treatment_statistics)
+
+    return EarlyStoppingTestStatistics(variant_statistics.control_statistics,
+                                       variant_statistics.treatment_statistics,
+                                       float(mu_x - mu_y),
+                                       dict([(p * 100, v) for p, v in zip([p1, p2], credible_interval_delta)]),
+                                       None, None, bool(stop))
```
Comment (shansfolder) [inline expan/core/early_stopping.py]: ```bool(stop)``` might be unnecessary.

File: expan/core/early_stopping.py [hunk 5]
```diff
@@ -283,30 +255,21 @@ def bayes_factor(x, y, distribution='normal', num_iters=25000, inference='sampli
...
-    elif distribution == 'poisson':
-        variance = np.nanmean(np.array(traces['delta']))
-        return traces['delta'] / np.sqrt(np.absolute(variance))
-    else:
-        raise ValueError("model " + distribution + " is not implemented.")
+    return EarlyStoppingTestStatistics(variant_statistics.control_statistics,
+                                       variant_statistics.treatment_statistics,
+                                       float(mu_x - mu_y),
+                                       dict([(p*100, v) for p, v in zip([p1, p2], credible_interval)]),
+                                       None, None, bool(stop))
```
Comment (shansfolder) [inline expan/core/early_stopping.py]: ```bool(stop)``` might be unnecessary.

**Review by shansfolder — COMMENTED** [review]
LGTM. Let's see if anything breaks when we have unit tests.

File: expan/core/early_stopping.py (hunk 1 above)
Comment (daryadedik) [reply expan/core/early_stopping.py]: oh, sure!

File: expan/core/early_stopping.py (hunk 2 above)
Comment (daryadedik) [reply expan/core/early_stopping.py]: will do the same!

**Comment by daryadedik** [comment]
3 unit tests are not passing probably because the previous problem with the precision.

File: tests/tests_core/test_early_stopping.py [hunk 6]
```diff
@@ -7,23 +7,20 @@
...
-        self.rand_s3 = np.random.poisson(lam=1.0, size=1000)
-        self.rand_s4 = np.random.poisson(lam=3.0, size=1000)
-        self.rand_s5 = np.random.normal (loc=0.0, size=1000)
-        self.rand_s6 = np.random.normal (loc=0.1, size=1000)
+        self.rand_s1 = list(np.random.normal (loc=0.0, size=1000))
+        self.rand_s2 = list(np.random.normal (loc=0.1, size=1000))
+        self.rand_s3 = list(np.random.poisson(lam=1.0, size=1000))
+        self.rand_s4 = list(np.random.poisson(lam=3.0, size=1000))
+        self.rand_s5 = list(np.random.normal (loc=0.0, size=1000))
+        self.rand_s6 = list(np.random.normal (loc=0.1, size=1000))
```
Comment (shansfolder) [inline tests/tests_core/test_early_stopping.py]: after the updated check for ```ndarray```, maybe we don't need the ```list()``` function here.

Comment (daryadedik) [reply tests/tests_core/test_early_stopping.py]: then all tests that test statistics will fail. I can create additional test to test the correct input type.

File: expan/core/early_stopping.py [hunk 7]
```diff
@@ -105,97 +107,63 @@ def group_sequential(x,
     else:
         stop = False
 
+    #TODO: why return frequentist confidence interval in bayesian method
```
Comment (daryadedik) [inline expan/core/early_stopping.py]: @shansfolder regarding this comment: actually this is frequentist early stopping, not bayesian, so I will remove it.

Comment (shansfolder) [reply expan/core/early_stopping.py]: Oh I see. Sorry lol.

**Pull request merged**